### PR TITLE
Store ABIs burrow side

### PR DESCRIPTION
**Pull request by seanyoung**

### CODE REVIEW

File: execution/evm/state.go [hunk 1]
```diff
@@ -182,7 +203,7 @@ func (st *State) CreateAccount(address crypto.Address) {
 	st.updateAccount(&acm.Account{Address: address})
 }
 
-func (st *State) InitCode(address crypto.Address, code []byte) {
+func (st *State) InitCode(address crypto.Address, parent *crypto.Address, code []byte) {
```
Comment (silasdavis) [inline execution/evm/state.go]: All usages of `InitCode` except one pass nil for the second argument. I tend to think this is a smell that indicates that we should introduces two methods and share implementation.

Can we introudce `InitChildCode` or similar that takes the parent argument, does the forebear bit, then calls `InitCode`?

We can also drop the unnecessary pointer then

File: execution/evm/state.go [hunk 2]
```diff
@@ -194,7 +215,60 @@ func (st *State) InitCode(address crypto.Address, code []byte) {
...
+	}
+
+	// If we have a list of ABIs for this contract, we also know what contract code it is allowed to create
+	// For compatibility with older contracts, allow any contract to be created if we have no mappings
+	if metamap != nil && len(metamap) > 0 {
+		found := false
+		for _, m := range metamap {
+			if bytes.Equal(codehash, m.CodeHash) {
+				found = true
+				break
```
Comment (silasdavis) [inline execution/evm/state.go]: Personally I'd rather see this broken out into a function where we can replace the `break`s with `return`s... I guess it's a matter of taste but I find that easier to follow with nesting etc.

File: deploy/compile/compilers.go [hunk 3]
```diff
@@ -46,21 +54,37 @@ type SolidityOutput struct {
...
 	Devdoc   json.RawMessage
 	Userdoc  json.RawMessage
 	Metadata string
+	// This is not present in the solidity output, but we add it ourselves
+	// This is map from DeployedBytecode to ABI. A Solidity contract can create any number
+	// of contracts, which have distinct ABIs. This is a map for the deployed code to abi,
+	// including the first contract itself.
+
+	// Note that libraries do not have ABIs. Also, the deployedbytecode does not match
+	// what Solidity tells use it will be.
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply deploy/compile/compilers.go]: This comment is out of date -- removed

File: deploy/def/client.go [hunk 4]
```diff
@@ -140,6 +141,22 @@ func (c *Client) GetAccount(address crypto.Address) (*acm.Account, error) {
 	return c.queryClient.GetAccount(ctx, &rpcquery.GetAccountParam{Address: address})
 }
 
+func (c *Client) GetAbiForAccount(address crypto.Address) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+	defer cancel()
+	abi, err := c.queryClient.GetAbi(ctx, &rpcquery.GetAbiParam{Address: address})
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply deploy/def/client.go]: Renamed to metadata now

File: deploy/def/client.go [hunk 5]
```diff
@@ -140,6 +141,22 @@ func (c *Client) GetAccount(address crypto.Address) (*acm.Account, error) {
...
+	if err != nil {
+		return "", err
+	}
+
+	return abi.Abi, nil
+}
+
+func (c *Client) GetAbi(abihash acmstate.AbiHash) (string, error) {
+	panic("not implemented")
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply deploy/def/client.go]: This required for Client to implement acmstate.Reader (used for proposals). Commented and explained now

File: deploy/def/client.go [hunk 6]
```diff
@@ -140,6 +141,22 @@ func (c *Client) GetAccount(address crypto.Address) (*acm.Account, error) {
...
+	defer cancel()
+	abi, err := c.queryClient.GetAbi(ctx, &rpcquery.GetAbiParam{Address: address})
+	if err != nil {
+		return "", err
+	}
+
+	return abi.Abi, nil
+}
+
+func (c *Client) GetAbi(abihash acmstate.AbiHash) (string, error) {
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply deploy/def/client.go]: It is consistent with what we do elswehere, no harm in documented that though. Added in comment

File: rpc/rpcquery/query_server.go [hunk 7]
```diff
@@ -59,6 +61,40 @@ func (qs *queryServer) GetAccount(ctx context.Context, param *GetAccountParam) (
 	return acc, err
 }
 
+func (qs *queryServer) GetAbi(ctx context.Context, param *GetAbiParam) (*AbiValue, error) {
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply rpc/rpcquery/query_server.go]: It is consistent with what we do elswehere, no harm in documented that though. Added in comment

File: protobuf/acm.proto [hunk 8]
```diff
@@ -25,4 +25,14 @@ message Account {
     bytes EVMCode = 5 [(gogoproto.customtype) = "Bytecode", (gogoproto.nullable) = false];
     permission.AccountPermissions Permissions = 6 [(gogoproto.nullable) = false];
     bytes WASMCode = 7 [(gogoproto.customtype) = "Bytecode", (gogoproto.jsontag) = ",omitempty", (gogoproto.nullable) = false];
+    bytes CodeHash = 8 [(gogoproto.customtype) = "github.com/hyperledger/burrow/binary.HexBytes", (gogoproto.nullable) = false, (gogoproto.jsontag) = "-"];
+    repeated MetaMap MetaMap = 9;
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply protobuf/acm.proto]: Now called ContractMeta

File: protobuf/acm.proto [hunk 9]
```diff
@@ -25,4 +25,14 @@ message Account {
...
+    bytes Forebear = 10 [(gogoproto.customtype) = "github.com/hyperledger/burrow/crypto.Address"];
+}
+
+message MetaMap {
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply protobuf/acm.proto]: ContractMeta as per slack discussion

File: protobuf/payload.proto [hunk 10]
```diff
@@ -70,6 +70,13 @@ message CallTx {
     bytes Data = 5 [(gogoproto.customtype) = "github.com/hyperledger/burrow/binary.HexBytes", (gogoproto.nullable) = false];
     // WASM bytecode
     bytes WASM = 6 [(gogoproto.customtype) = "github.com/hyperledger/burrow/binary.HexBytes", (gogoproto.nullable) = false, (gogoproto.jsontag)="tags,omitempty"];
+    // Set of contracts this code will deploy
+    repeated Abis Abis = 7;
+}
+
+message Abis {
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply protobuf/payload.proto]: Again ContractMeta

File: execution/evm/state.go (hunk 1 above)
Comment (seanyoung) [reply execution/evm/state.go]: Good point; fixed

File: execution/evm/state.go (hunk 2 above)
Comment (seanyoung) [reply execution/evm/state.go]: Good point. The code is shorter now!

**Comment by seanyoung** [comment]
app47 tests deployment without metadata.
burrow deploy --local-abi does not fetch ABIs from burrow

File: acm/acmstate/memory_state.go [hunk 11]
```diff
@@ -9,17 +9,19 @@ import (
 )
 
 type MemoryState struct {
-	Accounts map[crypto.Address]*acm.Account
-	Storage  map[crypto.Address]map[binary.Word256][]byte
+	Accounts  map[crypto.Address]*acm.Account
+	Storage   map[crypto.Address]map[binary.Word256][]byte
+	Metadatas map[MetadataHash]string
```
Comment (gregdhill) [inline acm/acmstate/memory_state.go]: `Sheep` or `Sheeps`

File: acm/acmstate/state.go [hunk 12]
```diff
@@ -45,6 +107,16 @@ type StorageIterable interface {
 	IterateStorage(address crypto.Address, consumer func(key binary.Word256, value []byte) error) (err error)
 }
 
+type MetadataGetter interface {
+	// Get an Metadata by its hash. This is content-addressed
+	GetMetadata(metahash MetadataHash) (string, error)
```
Comment (gregdhill) [inline acm/acmstate/state.go]: I don't think this ever returns an error...

File: deploy/def/client.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (silasdavis) [reply deploy/def/client.go]: Still not very satisfactory... It's not an acmstate.Reader then is it. Can we either implement it or have a new interface with just the methods we actually need?

**Comment by silasdavis** [comment]
Can we add `MetadataHash` to `MetadataParam` as an option in query client's `GetMetadata` then we can have the deploy client actually implement `acmstate.Reader` rather than panic. This may be useful anyway who knows. Should be easy to implement.

File: acm/acmstate/state.go (hunk 12 above)
Comment (seanyoung) [reply acm/acmstate/state.go]: That's a bad implementation.

File: acm/acmstate/memory_state.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (seanyoung) [reply acm/acmstate/memory_state.go]: Yes, that word is an abomination

File: deploy/def/client.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (seanyoung) [reply deploy/def/client.go]: Implemented as you suggested. I'd been mulling this over during the weekend and came to the same conclusion.

File: deploy/jobs/jobs_contracts.go [hunk 13]
```diff
@@ -424,40 +437,56 @@ func FormulateCallJob(call *def.Call, do *def.DeployArgs, deployScript *def.Play
 	call.Fee = FirstOf(call.Fee, do.DefaultFee)
 	call.Gas = FirstOf(call.Gas, do.DefaultGas)
 
+	// Get address (possibly via key)
+	address, err := client.GetKeyAddress(call.Destination, logger)
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply deploy/jobs/jobs_contracts.go]: I agree, renamed

**Comment by seanyoung** [comment]
I've added some comments about forebear in acm.proto

**Pull request merged**